File: backend/app/schemas/metrics.py
from datetime import datetime

from pydantic import BaseModel


class MetricsOrganizerOption(BaseModel):
    id: str
    name: str


class MetricsTotals(BaseModel):
    events: int
    sessions: int
    speaker_assignments: int
    total_capacity: int
    registered_attendees: int
    cancelled_registrations: int
    occupancy_rate: float
    cancellation_rate: float
    published_rate: float
    upcoming_events: int


class StatusTotal(BaseModel):
    status: str
    total: int


class MonthTotal(BaseModel):
    month: str
    total: int


class OrganizerPerformance(BaseModel):
    id: str
    name: str
    events: int
    registered: int
    total_capacity: int
    occupancy_rate: float


class SpeakerPerformance(BaseModel):
    id: str
    name: str
    sessions: int


class EventOccupancy(BaseModel):
    event_id: str
    event_name: str
    organizer_name: str
    capacity: int
    registered: int
    cancelled: int
    occupancy_rate: float


class SessionsPerEvent(BaseModel):
    event_id: str
    event_name: str
    sessions: int


class MetricsFilters(BaseModel):
    start_date: datetime | None = None
    end_date: datetime | None = None
    event_status: str | None = None
    organizer_id: str | None = None


class MetricsSummaryResponse(BaseModel):
    filters: MetricsFilters
    organizers: list[MetricsOrganizerOption]
    totals: MetricsTotals
    events_by_status: list[StatusTotal]
    sessions_by_status: list[StatusTotal]
    registrations_by_status: list[StatusTotal]
    events_by_month: list[MonthTotal]
    registrations_by_month: list[MonthTotal]
    top_organizers: list[OrganizerPerformance]
    top_speakers: list[SpeakerPerformance]
    event_occupancy: list[EventOccupancy]
    sessions_per_event: list[SessionsPerEvent]

File: frontend/src/types/metrics.ts
export type MetricsStatusTotal = {
  status: string;
  total: number;
};

export type MetricsMonthTotal = {
  month: string;
  total: number;
};

export type MetricsOrganizerOption = {
  id: string;
  name: string;
};

export type MetricsTotals = {
  events: number;
  sessions: number;
  speaker_assignments: number;
  total_capacity: number;
  registered_attendees: number;
  cancelled_registrations: number;
  occupancy_rate: number;
  cancellation_rate: number;
  published_rate: number;
  upcoming_events: number;
};

export type MetricsTopOrganizer = {
  id: string;
  name: string;
  events: number;
  registered: number;
  total_capacity: number;
  occupancy_rate: number;
};

export type MetricsTopSpeaker = {
  id: string;
  name: string;
  sessions: number;
};

export type MetricsEventOccupancy = {
  event_id: string;
  event_name: string;
  organizer_name: string;
  capacity: number;
  registered: number;
  cancelled: number;
  occupancy_rate: number;
};

export type MetricsSessionsPerEvent = {
  event_id: string;
  event_name: string;
  sessions: number;
};

export type MetricsSummaryResponse = {
  filters: {
    start_date: string | null;
    end_date: string | null;
    event_status: string | null;
    organizer_id: string | null;
  };
  organizers: MetricsOrganizerOption[];
  totals: MetricsTotals;
  events_by_status: MetricsStatusTotal[];
  sessions_by_status: MetricsStatusTotal[];
  registrations_by_status: MetricsStatusTotal[];
  events_by_month: MetricsMonthTotal[];
  registrations_by_month: MetricsMonthTotal[];
  top_organizers: MetricsTopOrganizer[];
  top_speakers: MetricsTopSpeaker[];
  event_occupancy: MetricsEventOccupancy[];
  sessions_per_event: MetricsSessionsPerEvent[];
};
